Adding mutual information loss

qiuhuaqi opened the pull request:
TODOs:

- [x] Add mutual information loss 
- [x] Add functionality to sample within a mask when computing MI

aschuh-hf commented on deepali/losses/functional.py:
```diff
@@ -636,6 +639,83 @@ def ssd_loss(
+        vmax: Optional[float] = None,
+        num_bins: int = 64,
+        sample_ratio: Optional[float] = None,
+        nmi: bool = True
```
Given that the function name includes a reference to "mutual information" (MI) already, I would suggest renaming this keyword argument to "normalized" or "normalize".

qiuhuaqi replied:
Agreed. Changing to `normalized`

aschuh-hf commented on deepali/losses/functional.py:
```diff
@@ -636,6 +639,83 @@ def ssd_loss(
+            if `None` mutual information is calculated using all voxels .
+        nmi: Calculate Normalised Mutual Information instead of Mutual Information if True.
+    """
+    input = masked_loss(input, mask=mask)
```
Wouldn't it be more appropriate to ensure that samples from which the intensity distribution is estimated only include voxels within the mask?

qiuhuaqi replied:
It would be more appropriate in theory. But if we have foreground masks with varying numbers of foreground voxels in a batch, sampling the same ratio of voxels means the sizes along the batch dimension would not be the same, which is not allowed in the tensor structure. 

there might be some hack around this issue within a batch. But if the tensor size changes between batches, it might lead to memory fragmentation issues in deep learning training, leading to higher GPU memory usage.

qiuhuaqi replied:
I can think of a perhaps naive hack around both these issues, which is to use the bounding box of the foreground mask padded to the same pre-defined size. Then we sample within this bounding box. But this seems a bit too hacky?

aschuh-hf replied:
Do you think `torch.randperm` could be replaced with `torch.multinomial(mask)` to achieve this?

That function is limited, however, by the size of the mask which cannot exceed 2 ** 24 grid element. Hence, we would need to adapt a custom implementation of inverse transform sampling. See below snippet copied from some other code where I sample random image patches for example. You might want to draw without replacement.

```
        # Indices of randomly sampled image patch ROI centers
        #
        # It is important to draw samples with replacement here. Otherwise, smaller ROIs
        # such as in particular diseased regions are undersampled even if their respective sampling
        # weight is considerably higher. Duplicate samples may be discarded afterwards again based on
        # the maximum ROI IoU. For this, we draw more samples than required (cf. num_samples above).
        #
        # As torch.multinomial() works only for at most 2^24 categories, use inverse transform sampling
        # if the number of candidate voxels with a sampling weight greater than zero is very large.
        if len(candidates) > 2 ** 24:
            weight_cdf = weight.astype(np.float64).flatten()
            weight_cdf = np.cumsum(weight_cdf / weight_cdf.sum())
            weight_cdf = np.divide(weight_cdf, weight_cdf[-1], out=weight_cdf)
            cdf_values = torch.rand(num_samples, generator=self.generator)
            imroi_index = np.searchsorted(weight_cdf, cdf_values.numpy())
            imroi_index = np.clip(imroi_index, 0, weight_cdf.size - 1, out=imroi_index)
        else:
            w = torch.from_numpy(weight.flatten()[candidates]).type(torch.float32)
            i = torch.multinomial(w, num_samples, generator=self.generator, replacement=True)
            imroi_index = candidates[i.numpy()]
```

aschuh-hf reviewed (commented):
Thanks, Huaqi, for contributing this! Looks good. I've only added minor suggestions. Most interestingly only the question in regards to how to handle a foreground mask. What do you think?

aschuh-hf commented:
Hey Huaqi, sorry I haven't yet had the time to review the changes you made. On first glance, this looks great!

After reading a paper that made reference to it, just wanted to drop a note that we could have also considered the [Gumbel-max trick](https://arxiv.org/abs/2110.01515) to sample the top N voxels within the mask without replacement. As an alternative to inverse transform sampling with replacement. No need to adopt this for this PR, just leaving this note here for reference later.

aschuh-hf commented on deepali/losses/functional.py:
```diff
@@ -638,6 +640,60 @@ def ssd_loss(
+    return sample_index
+
+
+def sample_in_mask(
```
What do you think about naming this function `rand_sample` and move it to `deepali.core.image`?

The function could be modified to take in one or more tensors to sample from with the same random set of indices.

```
def rand_sample(data: Union[Tensor, Tuple[Tensor, ...]], mask: Optional[Tensor] = None, ...) -> Union[Tensor, Tuple[Tensor, ...]]:
    ...

```

aschuh-hf commented on deepali/losses/functional.py:
```diff
@@ -638,6 +640,60 @@ def ssd_loss(
     return loss
 
 
+def sample_index_from_mask(mask: Tensor, num_samples: int):
```
Could we possibly move `sample_index_from_mask` to `deepali.core.random` and name it `multinomial`? As it is a wrapper around `torch.multinomial` which only extends its functionality to a larger number of categories (for `replacement=True`). We can extend this later also to include the sampling without replacement using the Gumbel-max trick.

I think we can leave the function as is without moving and renaming it for this PR and do this in a separate PR.

aschuh-hf reviewed (commented):
Hey Huaqi, I finally had a closer look and I think this mostly looks all well. The one thing I would ask prior to merging is to run the `black` code formatter.

The other minor items I would also be fine to leave them for later after merging this PR as is. I could also do some of this minor refactoring (i.e., move `sample_index_in_mask` and `sample_in_mask` to `deepali.core`).

qiuhuaqi commented:
hey Andreas, I think I've addressed most of the comments you left in your last review. 

I have modified the sampling function into `random_sample()` and `multinomial()` as you suggested. I agree these can be moved to `deepali.core` but I think I'll leave that to you so maybe you can make them useful for other use cases in this library. Of course I'm happy to help with that as well. 

otherwise I think this PR is ready to be merged?

aschuh-hf commented:
Great! Thanks for preparing those functions already for integration into core. Looks good!
